test/test_settings.c

```diff
@@ -0,0 +1,69 @@
+﻿#include <LCUI_Build.h>
+#include <LCUI/settings.h>
+#include <LCUI/main.h>
+#include "test.h"
+#include "libtest.h"
+
+static int settings_change_count = 0;
+
+static void on_settings_change(LCUI_SysEvent object, void *data)
+{
+	++settings_change_count;
+}
+
+static void test_default_settings(void)
+{
+	LCUI_SettingsRec settings;
+
+	LCUI_ResetSettings();
+	Settings_Init(&settings);
+
+	it_i("check default frame rate cap", settings.frame_rate_cap, 120);
+	it_i("check default parallel rendering threads",
+	     settings.parallel_rendering_threads, 4);
+	it_b("check default record profile", settings.record_profile, FALSE);
+	it_b("check default fps meter", settings.fps_meter, FALSE);
+	it_b("check default paint flashing", settings.paint_flashing, FALSE);
+}
+
+static void test_apply_settings(void)
+{
+	LCUI_SettingsRec settings;
+
+	LCUI_BindEvent(LCUI_SETTINGS_CHANGE, on_settings_change, NULL, NULL);
+
+	settings.frame_rate_cap = 60;
+	settings.parallel_rendering_threads = 2;
+	settings.record_profile = TRUE;
+	settings.fps_meter = TRUE;
+	settings.paint_flashing = TRUE;
+
+	LCUI_ApplySettings(&settings);
+
+	it_i("check frame rate cap", settings.frame_rate_cap, 60);
+	it_i("check parallel rendering threads",
+	     settings.parallel_rendering_threads, 3);
+	it_b("check record profile", settings.record_profile, TRUE);
+	it_b("check fps meter", settings.fps_meter, TRUE);
+	it_b("check paint flashing", settings.paint_flashing, TRUE);
+
+	it_i("check settings change count", settings_change_count, 1);
+
+	settings.frame_rate_cap = -1;
```

Comment:
I think some tests should be added to confirm whether settings.frame_rate_cap is really working, the test code is similar to this:

```c
void test_settings(void)
{
        ...
        describe("settings.frame_rate_cap", test_settings_frame_rate_cap);
}
...
void test_settings_frame_rate_cap(void)
{
...
        settings.frame_rate_cap = 30;
...
        LCUI_SetTimeout(1000, check_settings_frame_rate_cap, &settings.frame_rate_cap);
        LCUI_Main();

        settings.frame_rate_cap = 5;
        LCUI_SetTimeout(1000, check_settings_frame_rate_cap, &settings.frame_rate_cap);
        LCUI_Main();
}
...

void check_settings_frame_rate_cap(void *arg)
{
        int fps_limit = *((int*)arg);
        int fps = LCUI_GetFrameCount();
        char str[64];

        sprintf("str, should work when its value is %d", fps_limit);
        it_b(str, fps <= fps_limit && fps > fps_limit - 2);
        LCUI_Quit();
}
```

Reply:
I added the tests suggested, however I changed the accepted fps range to be (fps_limit - 10, fps_limit] rather than (fps_limit - 2, fps_limit]. The reason is that I noticed memcheck runs on Travis would sometimes be more than 5 frames less than the frame limit.